node/bindnode/infer.go

```diff
@@ -87,6 +104,78 @@ func fieldNameFromSchema(name string) string {
 	return strings.Title(name)
 }
 
+var defaultTypeSystem schema.TypeSystem
+
+func init() {
+	defaultTypeSystem.Init()
+
+	defaultTypeSystem.Accumulate(schemaTypeBool)
+	defaultTypeSystem.Accumulate(schemaTypeInt)
+	defaultTypeSystem.Accumulate(schemaTypeFloat)
+	defaultTypeSystem.Accumulate(schemaTypeString)
+	defaultTypeSystem.Accumulate(schemaTypeBytes)
+	defaultTypeSystem.Accumulate(schemaTypeLink)
+}
+
+// TODO: support IPLD maps and unions in inferSchema
+
+// TODO: support bringing your own TypeSystem?
+
+// TODO: we should probably avoid re-spawning the same types if the TypeSystem
+// has them, and test that that works as expected
+
 func inferSchema(typ reflect.Type) schema.Type {
-	panic("TODO")
+	switch typ.Kind() {
+	case reflect.Bool:
+		return schemaTypeBool
+	case reflect.Int64:
+		return schemaTypeInt
+	case reflect.Float64:
+		return schemaTypeFloat
+	case reflect.String:
+		return schemaTypeString
+	case reflect.Struct:
+		fieldsSchema := make([]schema.StructField, typ.NumField())
+		for i := range fieldsSchema {
+			field := typ.Field(i)
+			ftyp := field.Type
+			ftypSchema := inferSchema(ftyp)
+			fieldsSchema[i] = schema.SpawnStructField(
+				field.Name, // TODO: allow configuring the name with tags
+				ftypSchema.Name(),
+
+				// TODO: support nullable/optional with tags
+				false,
+				false,
+			)
+		}
+		name := schema.TypeName(typ.Name())
+		if name == "" {
+			panic("TODO: anonymous composite types")
+		}
```

Comment:
do you have a strategy for these anonymous types? data-dependent runtime panics are something it seems very preferable to avoid

Reply:
All these TODO panics will need to be fleshed out before we can call bindnode "ready for production use" for sure.

I left a comment further up with what I'm doing for anonymous slices and why. Structs are trickier, because I don't want to do `Struct_Foo_Bar_Baz` for `struct{ Foo; Bar; Baz }`. So I might end up just erroring in this scenario, or using incrementing struct names like `Struct_1`, `Struct_2`, etc.

Reply:
That said, data-dependent errors aren't particularly nice, either.

Thinking about it some more, this only changes behavior based on Go types, and nearly every bindnode user will be using statically declared Go types. So... I don't think these edge cases would surprise anyone, ever.